pkg/daemon/daemon.go

```diff
@@ -121,64 +154,78 @@ func (dn *Daemon) Run(stop <-chan struct{}) error {
-//    because of a machine reboot. validate the current machine state is the
-//    desired machine state. if we aren't try updating again. if we are, update
-//    the current state annotation accordingly.
-// 2. watch the desired config annotation, waiting for an update to be
-//    requested by the controller.
-// 3. if an update is requested by the controller, we assume that that means
-//    something changed and apply the desired config no matter what.
-// 4. the update function doesn't return right now, but at some point in the
-//    future if a reboot isn't required for an update it will. if it returns,
-//    validate the machine state and set the update to done.
-//
-// the only reason this function will return is if an error occurs. otherwise it
-// will keep trying to update the machine until it reboots.
-func (dn *Daemon) process() error {
-	for {
-		// validate machine state
-		isDesired, err := dn.isDesiredMachineState()
-		if err != nil {
-			return err
-		}
+// handleNodeUpdate is the handler for informer callbacks detecting node
+// changes. If an update is requested by the controller, we assume that
+// that means something changed and apply the desired config no matter what.
+// Also note that we only care about node updates, not creation or deletion.
+func (dn *Daemon) handleNodeUpdate(old, cur interface{}) {
+	node := cur.(*corev1.Node)
+
+	// First check if the node that was updated is this daemon's node
+	if (node.Name != dn.name) {
+		// The node that was changed was not ours
+		return
+	}
 
-		if isDesired {
-			// we got the machine state we wanted. set the update complete!
-			if err := dn.completeUpdate(); err != nil {
-				return err
-			}
+	// Then check we're not already in a degraded state.
+	if state, err := getNodeAnnotation(dn.kubeClient.CoreV1().Nodes(), dn.name, MachineConfigDaemonStateAnnotationKey); err != nil {
+		// try to set to degraded... because we failed to check if we're degraded
+		glog.Errorf("Marking degraded due to: %v", err)
+		setUpdateDegraded(dn.kubeClient.CoreV1().Nodes(), dn.name)
+		return
+	} else if state == MachineConfigDaemonStateDegraded {
+		// Just return since we want to continue sleeping
+		return
+	}
 
-			// now wait until we need another one.
-			glog.V(2).Infof("Watching for node annotation updates on %q", dn.name)
-			if err := waitUntilUpdate(dn.kubeClient.CoreV1().Nodes(), dn.name); err != nil {
-				return err
-			}
-		}
+	// Detect if there is an update
+	if (node.Annotations[DesiredMachineConfigAnnotationKey] == node.Annotations[CurrentMachineConfigAnnotationKey]) {
+		// No actual update to the config
+		return
+	}
 
-		// either the machine state isn't what we wanted and we should try
-		// again, or the machine state is what we wanted, and now another update
-		// is was triggered.
-		if err := dn.triggerUpdate(); err != nil {
-			return err
+	// The desired machine config has changed, trigger update
+	if err := dn.triggerUpdate(); err != nil {
```

Comment:
Do I understand correctly that there's nothing preventing this to run at the same time as the initial `triggerUpdate()`? Or do callbacks not actually occur until after `cache.WaitForCacheSync()`

Reply:
Good point, I haven't seen this cause an issue/race condition (even when there are repeated failed updates so it should have triggered), but let me double check.